File: ReactAndroid/src/main/java/com/facebook/react/views/image/ReactImageDownloadListener.java
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.react.views.image;

import android.graphics.Canvas;
import android.graphics.ColorFilter;
import android.graphics.PixelFormat;
import android.graphics.drawable.Animatable;
import android.graphics.drawable.Drawable;
import com.facebook.drawee.controller.ControllerListener;
import com.facebook.drawee.drawable.ForwardingDrawable;
import javax.annotation.Nullable;

public class ReactImageDownloadListener<INFO> extends ForwardingDrawable
    implements ControllerListener<INFO> {

  private static final int MAX_LEVEL = 10000;

  public ReactImageDownloadListener() {
    super(new EmptyDrawable());
  }

  public void onProgressChange(int loaded, int total) {}

  @Override
  protected boolean onLevelChange(int level) {
    onProgressChange(level, MAX_LEVEL);
    return super.onLevelChange(level);
  }

  @Override
  public void onSubmit(String id, Object callerContext) {}

  @Override
  public void onFinalImageSet(
      String id, @Nullable INFO imageInfo, @Nullable Animatable animatable) {}

  @Override
  public void onIntermediateImageSet(String id, @Nullable INFO imageInfo) {}

  @Override
  public void onIntermediateImageFailed(String id, Throwable throwable) {}

  @Override
  public void onFailure(String id, Throwable throwable) {}

  @Override
  public void onRelease(String id) {}

  /** A {@link Drawable} that renders nothing. */
  private static final class EmptyDrawable extends Drawable {

    @Override
    public void draw(Canvas canvas) {
      // Do nothing.
    }

    @Override
    public void setAlpha(int alpha) {
      // Do nothing.
    }

    @Override
    public void setColorFilter(ColorFilter colorFilter) {
      // Do nothing.
    }

    @Override
    public int getOpacity() {
      return PixelFormat.OPAQUE;
    }
  }
}

File: packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/image/ReactImageDownloadListener.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.react.views.image

import android.graphics.Canvas
import android.graphics.ColorFilter
import android.graphics.PixelFormat
import android.graphics.drawable.Animatable
import android.graphics.drawable.Drawable
import com.facebook.drawee.controller.ControllerListener
import com.facebook.drawee.drawable.ForwardingDrawable

internal open class ReactImageDownloadListener<INFO> :
    ForwardingDrawable(EmptyDrawable()), ControllerListener<INFO> {
  open fun onProgressChange(loaded: Int, total: Int) = Unit

  override fun onLevelChange(level: Int): Boolean {
    onProgressChange(level, MAX_LEVEL)
    return super.onLevelChange(level)
  }

  override fun onSubmit(id: String, callerContext: Any?) = Unit

  override fun onFinalImageSet(id: String, imageInfo: INFO?, animatable: Animatable?) = Unit

  override fun onIntermediateImageSet(id: String, imageInfo: INFO?) = Unit

  override fun onIntermediateImageFailed(id: String, throwable: Throwable) = Unit

  override fun onFailure(id: String, throwable: Throwable) = Unit

  override fun onRelease(id: String) = Unit

  /** A [Drawable] that renders nothing. */
  private class EmptyDrawable : Drawable() {
    override fun draw(canvas: Canvas) = Unit

    override fun setAlpha(alpha: Int) = Unit

    override fun setColorFilter(colorFilter: ColorFilter?) = Unit

    override fun getOpacity(): Int = PixelFormat.OPAQUE
  }

  companion object {
    private const val MAX_LEVEL = 10000
  }
}
